pallets/stream-payment/src/tests.rs

```diff
@@ -0,0 +1,1213 @@
+
+            let new_config = StreamConfig {
+                rate: 101,
+                ..open_stream.config
+            };
+
+            assert_ok!(StreamPayment::request_change(
+                RuntimeOrigin::signed(ALICE),
+                0,
+                ChangeKind::Suggestion,
+                new_config,
+                None,
+            ),);
+
+            assert_event_emitted!(Event::<Runtime>::StreamConfigChanged {
+                stream_id: 0,
+                old_config: open_stream.config,
+                new_config,
+                deposit_change: None,
+            });
+        })
+    }
+
+    #[test]
+    fn target_can_immediately_decrease_rate() {
+        ExtBuilder::default().build().execute_with(|| {
+            let open_stream = OpenStream::default();
+            assert_ok!(open_stream.call());
+
+            let new_config = StreamConfig {
+                rate: 99,
+                ..open_stream.config
+            };
+            assert_ok!(StreamPayment::request_change(
+                RuntimeOrigin::signed(BOB),
+                0,
+                ChangeKind::Suggestion,
+                new_config,
+                None,
+            ),);
+
+            assert_event_emitted!(Event::<Runtime>::StreamConfigChanged {
+                stream_id: 0,
+                old_config: open_stream.config,
+                new_config,
+                deposit_change: None,
+            });
+        })
+    }
+
+    fn source_can_immediately_change_deposit(change: DepositChange<Balance>) {
+        ExtBuilder::default().build().execute_with(|| {
+            let open_stream = OpenStream::default();
+            assert_ok!(open_stream.call());
+
+            assert_ok!(StreamPayment::request_change(
+                RuntimeOrigin::signed(ALICE),
+                0,
+                ChangeKind::Suggestion,
+                open_stream.config,
+                Some(change),
+            ));
+
+            assert_event_emitted!(Event::<Runtime>::StreamConfigChanged {
+                stream_id: 0,
+                old_config: open_stream.config,
+                new_config: open_stream.config,
+                deposit_change: Some(change),
+            });
+
+            assert_balance_change!(-, ALICE, match change {
+                DepositChange::Absolute(amount) => amount,
+                DepositChange::Increase(amount) => open_stream.deposit + amount,
+                DepositChange::Decrease(amount) => open_stream.deposit - amount,
+            });
+        })
+    }
+
+    #[test]
+    fn source_can_immediately_change_deposit_absolute() {
+        source_can_immediately_change_deposit(DepositChange::Absolute(100))
+    }
+
+    #[test]
+    fn source_can_immediately_increase_deposit() {
+        source_can_immediately_change_deposit(DepositChange::Increase(100))
+    }
+
+    #[test]
+    fn source_can_immediately_decrease_deposit() {
+        source_can_immediately_change_deposit(DepositChange::Decrease(100))
```

Comment:
I think it's ok, sometimes source might have put too much in the deposit. The target as well can close the stream if the amount is not being paid, so I think it's ok

Reply:
That's indeed the idea. Since the source can close the stream at any time and recover unspent deposit, I think we can allow to decrease the deposit without closing the stream. One could first deposit a big amount because they want a long-term service, but at some point temporarily need some funds for something else. What matters for the target is that past time is guaranteed to be paid, and by looking at the events they can compute up to which time the stream is funded.